### ISSUE TITLE
Implemented Suggestion: polynomial to rational form. 

### ISSUE THREAD

**Issue by paulobuchsbaum** [josdejong/mathjs#926]
I've implemented a solution related to general polynomial expressions, that it is a subset of valid expressions in `math.js`.

I've tested in some different situations.

Maybe it can be interesting for some users.

First, I've made a recursive expand function only for expressions with exponent. This cannot be solved with  simplification rules.

```
//*******************************************************
//                       expandPower
// Expand recursively a tree "node" for handling with expressions with exponents
//  (it's not for constants, symbols or functions with exponents)
//
// Uses expandPower just with NODE parameter. It's an additional procedure 
// in simplify rational polynomials 
//
//*******************************************************
function expandPower(node,parent,indParent) {
var i, val, no, tp, does;
tp = node.type; 
var internal = (arguments.length>1)   // TRUE in internal calls

if (tp==='OperatorNode') { 
  does = false;
  if (node.op==="^")  {   // First operator: Parenthesis or UnaryMinus
    if ( ( node.args[0].type==="ParenthesisNode" ||  
           node.args[0].type==="OperatorNode" ) 
            && (node.args[1].type==="ConstantNode") )  {   // Second operator: Constant
        val = parseFloat(node.args[1].value);
        does = (val>=2 && Number.isInteger(val));  // Just for integer exponent > 2
    }
  } 

  if (does)  {

//Before:
//            operator A --> Subtree
// parent pow 
//            constant
//

     if (val>2)     {  // Exponent > 2, 

//AFTER:  (exponent > 2)
//                 operator A --> Subtree
// parent    * 
//                      deep clone (operator A --> Subtree
//                 pow     
//                     constant - 1
//
       var nEsqTopo = node.args[0];  
       var nDirTopo = new m.expression.node.OperatorNode('^', 'pow',
              [node.args[0].cloneDeep(),new m.expression.node.ConstantNode(val-1)]);
       node = new m.expression.node.OperatorNode("*",
                                         "multiply", [nEsqTopo, nDirTopo]);
     } else   // Expo = 2 - no power

//AFTER:  (exponent =  2)
//                operator A --> Subtree
// parent    oper 
//                deep clone (operator A --> Subtree)    
//                            
       node = new m.expression.node.OperatorNode("*", "multiply",
                 [node.args[0], node.args[0].cloneDeep()]);

       // Change parent references in internal recursive calls
     if (internal) 
        if (indParent==="content")
           parent.content = node;
        else
           parent.args[indParent] = node
  }
}
// Recursion in chidren
  if (tp==="ParenthesisNode" ) {
     if (!! node.content)
       expandPower(node.content,node,"content");  
     for (i=0;i<node.args.length;i++)  
       expandPower(node.args[i],node,i);
  } else if (tp!=='ConstantNode' && tp!=='SymbolNode')  
    for (i=0;i<node.args.length;i++)  
      expandPower(node.args[i],node,i);
    
  
 if (! internal )   // return the root node
     return node
}  // End expandPower
```

After I've made a code chunk for generate rules in 2 stages. 

```
/*******************************************************
                    rulesRationalize
Return a rule set to rationalize an polynomial expression 
********************************************************/
function rulesRationalize() {
var rules;  
var defRules = m.simplify.rules; // Standard simplify rules
var oldRules = defRules.slice() // clone default rules to change
var setRules={};   // rules set in 2 steps. 

// Delete the default rules that don't match with rationalize polynomials
 oldRules.splice(23,1); 
 oldRules.splice(15,4);         
 oldRules.splice(12,2); 
 oldRules.splice(1,9);  
 
rules = [
  { l: '(-n1)/(-n2)', r: 'n1/n2' },  // Unary division
  { l: '(-n1)*(-n2)', r: 'n1*n2' },  // Unary multiplication
  { l: "n1--n2", r:"n1+n2"},        // "--" elimination
  { l: 'n1-n2', r:'n1+(-n2)'} ,      // Subtraction turn into add with unáry minus    
  { l:'(n1+n2)*n3', r:'(n1*n3 + n2*n3)' },     // Distributive 1 
  { l:'n1*(n2+n3)', r:'(n1*n2+n1*n3)' },       // Distributive 2 
  { l: 'c1*n + c2*n', r:'(c1+c2)*n'} ,       // Joining constants
  { l: '-v*-c', r:'c*v'} ,          // Inversion constant and variable 1
  { l: '-v*c', r:'-c*v'} ,          // Inversion constant and variable 2
  { l: 'v*-c', r:'-c*v'} ,          // Inversion constant and variable 3
  { l: 'v*c', r:'c*v'} ,            // Inversion constant and variable 4
  { l: '-(-n1*n2)', r:'(n1*n2)'} ,  // Unary propagation
  { l: '-(n1*n2)', r:'(-n1*n2)'} ,  // Unary propagation
  { l: '-(-n1+n2)', r:'(n1-n2)'} ,  // Unary propagation
  { l: '-(n1+n2)', r:'(-n1-n2)'} ,  // Unary propagation
  { l: '(n1^n2)^n3', r:'(n1^(n2*n3))'} ,  // Power to Power
  { l: '-(-n1/n2)', r:'(n1/n2)'} ,   // Division and Unary
  { l: '-(n1/n2)', r:'(-n1/n2)'} ,   // Divisao and Unary
  { l:'(n1/(n2/n3))', r:'((n1*n3)/n2)'} , // Division simplification
  { l:'(n1/n2/n3)', r:'(n1/(n2*n3))' } ,  // Division simplification
  { l:'(n1/n2 + n3/n4)', r:'((n1*n4 + n3*n2)/(n2*n4))' },  // Sum of fractions
  { l:'(n1/n2 + n3)', r:'((n1 + n3*n2)/n2)' }, // Sum fraction with number 1
  { l:'(n1 + n2/n3)', r:'((n1*n3 + n2)/n3)' }   // Sum fraction with number 1
  ];
setRules.rules1=oldRules.concat(rules);  // First rule set

// Second rule set. 
// There is no aggregate expression with parentesis,
// but the only variable can be scattered. 
setRules.rules2=[    defRules[0],
 { l: 'n*-n', r: '-n^2' },                // Joining multiply with power 1
 { l: 'n*n', r: 'n^2' },                  // Joining multiply with power 2
 { l: 'n*-n^n1', r: '-n^(n1+1)' },        // Joining multiply with power 3
 { l: 'n*n^n1', r: 'n^(n1+1)' },          // Joining multiply with power 4
 { l: 'n^n1*-n^n2', r: '-n^(n1+n2)' },    // Joining multiply with power 5
 { l: 'n^n1*n^n2', r: 'n^(n1+n2)' },      // Joining multiply with power 6
 { l: 'n^n1*-n', r: '-n^(n1+1)' },        // Joining multiply with power 7
 { l: 'n^n1*n', r: 'n^(n1+1)' },          // Joining multiply with power 8
 { l: 'n^n1/-n', r: '-n^(n1-1)' },        // Joining multiply with power 8
 { l: 'n^n1/n', r: 'n^(n1-1)' },          // Joining division with power 1
 { l: 'n/-n^n1', r: '-n^(1-n1)' },        // Joining division with power 2
 { l: 'n/n^n1', r: 'n^(1-n1)' },          // Joining division with power 3
 { l: 'n^n1/-n^n2', r: 'n^(n1-n2)' },     // Joining division with power 4
 { l: 'n^n1/n^n2', r: 'n^(n1-n2)' },      // Joining division with power 5
 { l: 'n1+(-n2*n3)', r: 'n1-n2*n3' },     // Solving useless parenthesis 1 
 { l: 'n+(-c)', r: 'n-c' },               // Solving useless unary 1 
 { l: 'v*(-c)', r: '-c*v' },              // Solving useless unary 2 
 { l: 'v*c', r: 'c*v' },                  // inversion constant with variable
 { l: '(n1^n2)^n3', r:'(n1^(n2*n3))'},    // Power to Power
 defRules[14]
 ];                    
 return setRules;
} // End rulesRationalize
```

Finally the  function for the final user, that uses the 2 above function:

```
/*******************************************************
                    rational
Transform a polynomial expression in rational fraction.
ScopeAdd is an optional scope. 
********************************************************/
function rational(expr, scopeAdd) {
expr=expr.toLowerCase();                     // Uses lower case for an easier check. 

var exprint = "(?:[-()^+*]|\\d)+" ;           // Integer expression
var numpatt = "\\d+\\.?\\d*(?:e-?\\d{1,2})?"  // Define floating number

var escopo = {e:2.718281828459045,pi:3.141592653589793};  
if (arguments.length>1) {   // Add aditional scope. 
  escopo.concat(scopeAdd);  
}

 // Resolves any variables and functions with all defined parameters
var node=m.simplify(expr,escopo); 
expr=node.toString();     

   // define identifier (a little bit different from MathJs)
var idsearch=new RegExp("([a-z_]\\w*)","g");      
var res = idsearch.exec(expr);    // Search for the first identifier
if (! res)
   return "Expression doesn't contains any variable."; 

var varpatt = res[0];  // The only variable

     // polinomial expression with one variable(not yet simplified)
var pattpoly =new RegExp("^(?:" + numpatt + "|\\s+|" +
                       varpatt + "|\\^\\s*" + exprint + "|[-(/)+*])+$","gim");
if (! pattpoly.exec(expr))      // Is expression polinomial with one variable? 
  return 'Expression is not a polynomial with one variable';

var setRules = rulesRationalize();
node=expandPower(node);
node = m.simplify(node,setRules.rules1);     
node = m.simplify(node,setRules.rules2);
return node;
}
```

One simple test can be seen here:

```
expr = '2x/((2x-1)/(3x+2))-5x/((3x+4)/(2x^2-5))+3' 
console.log("Before: ",expr);
node = rational(expr);
if (typeof node==="string") 
  console.log(node)   
else 
  console.log("After: ",node.toString().replace(/ /g,''));  
console.log("...........................................");
expr = '(2x-1)^4'
console.log("Before: ",expr);
node = rational(expr);
if (typeof node==="string") 
  console.log(node)   
else 
  console.log("After: ",node.toString().replace(/ /g,''));  
return;
```

And the results:

> Before:  2x/((2x-1)/(3x+2))-5x/((3x+4)/(2x^2-5))+3     
> After:  (104*x^2-20*x^4-12-25*x-9*x+40*x+18*x^3+10*x^3)/(6*x^2-3*x-4+8*x)     
> ...........................................
> Before:  (2x-1)^4     
> After:  -8*x-32*x^3+24*x^2+16*x^4+1     

I don't solve the final simplification that means join the coefficients from the same exponent. It cannot be made by simplification rules, but it is not important and easy to make it with regex support.

**Comment by josdejong** [comment]
Nice!! So this is the followup of #920 right? Can you work these two new functions in a pull request?

**Comment by paulobuchsbaum** [comment]
Yes, of course.  It would be an honor for me, @josdejong 

As you can see, I've done pretty much everything in above code. I guess that it would be enough to make the test suite more robust and someone could  help me to integrate it into the mathjs factory environment and the GitHub protocols, because I have no experience in this field. Even the "pull request" term is new for me and I need to search for more information.

**Comment by josdejong** [comment]
cool! There is a lot of documentation on internet on how pull requests with git(hub) work, and if something is not clear I can help you further.

**Comment by paulobuchsbaum** [comment]
great! let's do it

**Comment by josdejong** [comment]
:muscle:

**Comment by paulobuchsbaum** [comment]
@josdejong, in my tests about rationalizing expressions, I've purposely have tried some more complex polynomial expressions like

>   x/(1-x)/(x-2)/(x-3)/(x-4) + 2x/ ( (1-2x)/(2-3x) )/ ((3-4x)/(4-5x) )

 I've come across a infinite loop (stuck) in simplify function because sucessives divides rules colides with divide distribution  rules for sum and subtraction. 

I got over this problem, dividing the rules in 3 sets.

However, that complication, even if  not inevitable, it is inherent to the algorithm  used in the simplification, that applyes all the rules until no change. 

It is happen because it's possible nested sucessive divisions inside nested fractions sums and subtraction  and vice versa.
 
So, a simple way to overcome this problem is creating an optional timeout parameter (in seconds) in simplify method that fires an exception in order to be possible to issue a message "expression too complex" in rationalize method. 

 I know that it do not would be hard because it's just measuring the time and test the time limit before running all the simplification rules once again.

**Comment by firepick1** [comment]
Oh wow. I see what you mean about the visited[str]. That's got a hard-wired assumption about convergence. That's the existing implementation and as such I'm a little nervous about tackling a fix, since the looping is a core design assumption. I'm also not too excited about a time-limit, since it is very "un-mathy". Not sure what to recommend here...

**Comment by paulobuchsbaum** [comment]
I understand it. 

It's impossible to avoid infinite loop in my concept because the heuristic of simplification rules implies this. 
It's a metaphor of the universe inside a nutshell. I need to handle with potentially multilevel of parenthesis, fractions sums and sucessive divides.

One way that I can try is successively run rules in stages and start it again, if it is necessary, until no more changes. It's impossible mix all rules because one rule result can generate a cause for another rule and vice versa. So a need to separate rules in different chunks. 

The only point is that if the user puts a sufficiently intricate  expression, maybe he will wait for a long time.

**Comment by paulobuchsbaum** [comment]
It works! 

I've separated distribution division rules from successive division rules. Ant it is enough!  

If someone crazy create a very intricate and nested function, it last a lot to run, but there is no infinite loop.

I've divided the rules in 4 groups:

> 1 = FirstRules.      
> s = Successive Division Rules     
```
 [ { l:'(n1/(n2/n3))', r:'((n1*n3)/n2)'} ,     // sucessive divisions I
  { l:'(n1/n2/n3)', r:'(n1/(n2*n3))' } ]      // sucessive divisions II 
```    
> d = Distribution Division Rules            
```
[  { l:'(n1/n2 + n3/n4)', r:'((n1*n4 + n3*n2)/(n2*n4))' },  // Sum of fractions
  { l:'(n1/n2 + n3)', r:'((n1 + n3*n2)/n2)' },    // Sum fraction with number 1
  { l:'(n1 + n2/n3)', r:'((n1*n3 + n2)/n3)' }  ];  // Sum fraction with number 2
```
> f = Final Rules (elements already atomized)         
 

I ran

```
rules 1 + s   // first rules + sucessive division
loop
  rules d   // distribution division
  if don't change break
  rules s  // successive division
  if don't change break
end loop
If there is ever any change
   rules 1   // for rearrange after division handling
rules f   // final rules
```
So, 

>   x/(1-x)/(x-2)/(x-3)/(x-4) + 2x/ ( (1-2x)/(2-3x) )/ ((3-4x)/(4-5x) )

results in

>   (-30x^7 +344x^6 -1506x^5 +3200x^4 -4267x^3 +1847x^2 -381x)/     
>   (-8x^6 +90x^5 -383x^4   +780x^3 -797x^2 +390x -72)      
 
However, if I run

```
rules 1 + s + d  // first rules + sucessive division + distribution division
rules f   // final rules
```    

We have a "gorgeous" infinite loop, even the rules alone are completely valid.

**Comment by firepick1** [comment]
Somehow, only in mathjs would I expect to hear about "gorgeous infinite loops". 😱 😅

**Comment by paulobuchsbaum** [comment]
Yes! Because it was initially surprising for me how somehow elegant and valid rules conjugated produce a bomb.

After I have done some manual simulations, I have found that these rules together end up alternating between digging and covering a hole, which doesn't have a happy end.

**Comment by josdejong** [comment]
So what would be a good solution to deal with infinite loops? Maybe we can simply put a safety maximum on the number of loops. Or maybe we can somehow keep track on which rules are applied to a specific part of the expression, and stop when we are about to apply the same rule again (ie. detect an infinite loop).

**Comment by paulobuchsbaum** [comment]
@josdejong, I've solved it in [pull request](https://github.com/josdejong/mathjs/pull/931). 

I've separated my  rules in sets and part of them I've run inside a loop, where I've checked myself if was some change for each run. If there are no changes, I exit my loop. 

It's curious because when I ran sucessive **divide rules** and **sum/subtraction fractions rules** together, there was a infinite loop, but when I ran separetely, I overcame the infinite loop. 

However, there is no protection for the general simplify user. Perfectly harmless rules can make `simplify` enter in a infinite loop and it was bad for the final user. 

>    Sorry for making a pull request in` master` branch. It was a mistake.      
>    I did not pay attention to the option to select the branch.        
>    When I've saw the pull request in master, it's too late to go back.
